contrib/mesos/pkg/scheduler/podtask/procurement.go

```diff
@@ -50,77 +62,212 @@ type AllOrNothingProcurement []Procurement
+		if mapping, err := t.mapper.Map(t, offer); err != nil {
+			return err
+		} else {
+			ports := []Port{}
+			for _, entry := range mapping {
+				ports = append(ports, Port{
+					Port: entry.OfferPort,
+					Role: entry.Role,
+				})
+			}
+			spec.PortMap = mapping
+			spec.Resources = append(spec.Resources, portRangeResources(ports)...)
+		}
+		return nil
+	})
 }
 
-func (r *RequirePodResources) Procure(t *T, offer *mesos.Offer) error {
-	// write resource limits into the pod spec which is transferred to the executor. From here
-	// on we can expect that the pod spec of a task has proper limits for CPU and memory.
-	// TODO(sttts): For a later separation of the kubelet and the executor also patch the pod on the apiserver
-	// TODO(sttts): fall back to requested resources if resource limit cannot be fulfilled by the offer
-	// TODO(jdef): changing the state of t.Pod here feels dirty, especially since we don't use a kosher
-	// method to clone the api.Pod state in T.Clone(). This needs some love.
-	_, cpuLimit, _, err := mresource.LimitPodCPU(&t.Pod, r.defaultContainerCPULimit)
-	if err != nil {
-		return err
-	}
+// NewExecutorResourceProcurer returns a Procurement procuring executor resources
+// If a given offer has no executor IDs set, the given prototype executor resources are considered for procurement.
+// If a given offer has one executor ID set, only pod resources are being procured.
+// An offer with more than one executor ID implies an invariant violation and the first executor ID is being considered.
+func NewExecutorResourceProcurer(resources []*mesos.Resource, registry executorinfo.Registry) Procurement {
+	return ProcurementFunc(func(t *T, offer *mesos.Offer, _ *api.Node, spec *Spec) error {
+		eids := len(offer.GetExecutorIds())
+		switch {
+		case eids == 0:
+			wantedCpus := sumResources(filterResources(resources, isScalar, hasName("cpus")))
+			wantedMem := sumResources(filterResources(resources, isScalar, hasName("mem")))
 
-	_, memLimit, _, err := mresource.LimitPodMem(&t.Pod, r.defaultContainerMemLimit)
-	if err != nil {
-		return err
-	}
+			procuredCpu, remaining := procureRoleResources("cpus", wantedCpus, t.allowedRoles, offer.GetResources())
+			if procuredCpu == nil {
+				return fmt.Errorf("not enough cpu resources for executor: want=%v", wantedCpus)
+			}
 
-	log.V(3).Infof("Recording offer(s) %s/%s against pod %v: cpu: %.2f, mem: %.2f MB", offer.Id, t.Pod.Namespace, t.Pod.Name, cpuLimit, memLimit)
+			procuredMem, remaining := procureRoleResources("mem", wantedMem, t.allowedRoles, remaining)
+			if procuredMem == nil {
+				return fmt.Errorf("not enough mem resources for executor: want=%v", wantedMem)
+			}
 
-	t.Spec.CPU = cpuLimit
-	t.Spec.Memory = memLimit
+			offer.Resources = remaining
+			spec.Executor = registry.New(offer.GetHostname(), append(procuredCpu, procuredMem...))
+			return nil
 
-	return nil
+		case eids > 1:
+			log.Errorf("got %d executor IDs in offer, want 1 (procuring first)", eids)
```

Comment:
Good idea. This should never happen. Much more important than this error: are we sure that this can never happen? Can there be any races when launching tasks?